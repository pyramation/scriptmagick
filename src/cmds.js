export const maskImage = ({ pathToMaskLayer, pathToObjectLayer, outFile }) => {
  return `convert ${pathToMaskLayer} ${pathToObjectLayer} -alpha on -compose CopyOpacity -composite ${outFile}`;
};

export const composeImages = ({ layers, outFile }) => {
  for (let i = 0; i < layers.length; i++) {
    const layer = layers[i];
    if (!layer) {
      console.log(JSON.stringify(layers, null, 4));
      throw new Error('missing a layer!');
    }
  }

  const pages = layers
    .reverse()
    .map((l) => `-page +0+0 ${l}`)
    .join(' ');
  return `convert ${pages} -layers flatten ${outFile}`;
};

export const replaceColor = ({
  inFile,
  find,
  replace,
  outFile,
  fuzz = '24%'
}) => {
  return [
    'convert',
    inFile,
    `-fuzz ${fuzz}`,
    `-fill "${replace}"`,
    `-opaque "${find}"`,
    outFile
  ].join(' ');
};
