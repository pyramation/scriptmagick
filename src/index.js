export * from './klass';
export * from './cmds';
