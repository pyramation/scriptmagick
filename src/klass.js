import * as shell from 'shelljs';
import { resolve, join } from 'path';
import { sync as mkdirp } from 'mkdirp';
import { sync as rimraf } from 'rimraf';

import * as cmds from './cmds';

export class ScriptMagick {
  constructor({ outDir }) {
    this.outDir = outDir;
    this.count = 0;
    mkdirp(outDir);
  }

  execCmd(cmd) {
    console.log(cmd);
    console.log('');
    shell.exec(cmd);
  }

  getIntermediateFile() {
    const c = this.count++ + '';
    const name = 'intermediate-' + c.padStart(8, '0') + '.png';
    return resolve(join(this.outDir, name));
  }

  maskImage({ pathToMaskLayer, pathToObjectLayer, outFile }) {
    if (!outFile) {
      outFile = this.getIntermediateFile();
    }
    const cmd = cmds.maskImage({ pathToMaskLayer, pathToObjectLayer, outFile });
    this.execCmd(cmd);
    return outFile;
  }

  composeImages({ layers, outFile }) {
    const cmd = cmds.composeImages({ layers, outFile });
    this.execCmd(cmd);
    return outFile;
  }

  replaceColor({ inFile, find, replace, outFile, fuzz = '24%' }) {
    const cmd = cmds.replaceColor({ inFile, find, replace, outFile, fuzz });
    this.execCmd(cmd);
  }
  cleanup() {
    rimraf(resolve(join(this.outDir, 'intermediate-*.png')));
  }
}
