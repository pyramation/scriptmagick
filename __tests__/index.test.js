import { maskImage, composeImages, replaceColor } from '../src';
import cases from 'jest-in-case';

cases(
  'maskImage',
  (options) => {
    expect(maskImage(options)).toMatchSnapshot();
  },
  [{ pathToMaskLayer: 'mask', pathToObjectLayer: 'obj', outFile: 'out' }]
);

cases(
  'composeImages',
  (options) => {
    expect(composeImages(options)).toMatchSnapshot();
  },
  [{ layers: ['a', 'b', 'c', 'd'], outFile: 'out' }]
);

cases(
  'replaceColor',
  (options) => {
    expect(replaceColor(options)).toMatchSnapshot();
  },
  [
    {
      inFile: 'in',
      find: 'find',
      replace: 'replace',
      outFile: 'out',
      fuzz: '42%'
    }
  ]
);
