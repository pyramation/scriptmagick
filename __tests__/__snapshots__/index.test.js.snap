// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composeImages case: 1 1`] = `"convert -page +0+0 d -page +0+0 c -page +0+0 b -page +0+0 a -layers flatten out"`;

exports[`maskImage case: 1 1`] = `"convert mask obj -alpha on -compose CopyOpacity -composite out"`;

exports[`replaceColor case: 1 1`] = `"convert in -fuzz 42% -fill \\"replace\\" -opaque \\"find\\" out"`;
